File: scripts/ocr_macos.m
#import <Foundation/Foundation.h>
#import <Vision/Vision.h>
#import <ImageIO/ImageIO.h>

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        if (argc != 2) {
            fprintf(stderr, "uso: bft_ocr_macos imagem.png\n");
            return 2;
        }
        NSString *caminho = [NSString stringWithUTF8String:argv[1]];
        NSURL *url = [NSURL fileURLWithPath:caminho];
        CGImageSourceRef fonte = CGImageSourceCreateWithURL((__bridge CFURLRef)url, NULL);
        if (!fonte) {
            fprintf(stderr, "imagem inválida\n");
            return 3;
        }
        CGImageRef imagem = CGImageSourceCreateImageAtIndex(fonte, 0, NULL);
        CFRelease(fonte);
        if (!imagem) {
            fprintf(stderr, "imagem inválida\n");
            return 3;
        }

        VNRecognizeTextRequest *requisicao = [[VNRecognizeTextRequest alloc] init];
        requisicao.recognitionLevel = VNRequestTextRecognitionLevelAccurate;
        requisicao.usesLanguageCorrection = NO;
        VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCGImage:imagem options:@{}];
        NSError *erro = nil;
        BOOL ok = [handler performRequests:@[requisicao] error:&erro];
        CGImageRelease(imagem);
        if (!ok) {
            fprintf(stderr, "OCR falhou: %s\n", erro.localizedDescription.UTF8String);
            return 4;
        }
        for (VNRecognizedTextObservation *observacao in requisicao.results) {
            VNRecognizedText *candidato = [[observacao topCandidates:1] firstObject];
            if (candidato) {
                printf("%s\n", candidato.string.UTF8String);
            }
        }
    }
    return 0;
}

File: scripts/ocr_macos.swift
import Foundation
import Vision
import ImageIO

guard CommandLine.arguments.count == 2 else {
    fputs("uso: ocr_macos.swift imagem.png\n", stderr)
    exit(2)
}

let url = URL(fileURLWithPath: CommandLine.arguments[1])
guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
      let imagem = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
    fputs("imagem inválida\n", stderr)
    exit(3)
}

let requisicao = VNRecognizeTextRequest()
requisicao.recognitionLevel = .accurate
requisicao.usesLanguageCorrection = false
requisicao.recognitionLanguages = ["pt-BR", "en-US"]

do {
    try VNImageRequestHandler(cgImage: imagem, options: [:]).perform([requisicao])
    let observacoes = (requisicao.results ?? []).sorted {
        if abs($0.boundingBox.midY - $1.boundingBox.midY) > 0.01 {
            return $0.boundingBox.midY > $1.boundingBox.midY
        }
        return $0.boundingBox.minX < $1.boundingBox.minX
    }
    for observacao in observacoes {
        if let texto = observacao.topCandidates(1).first?.string {
            print(texto.replacingOccurrences(of: "\n", with: " "))
        }
    }
} catch {
    fputs("OCR falhou: \(error)\n", stderr)
    exit(4)
}
